drivers/dac/dac_ad56x1.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/dac.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+LOG_MODULE_REGISTER(dac_ad56x1, CONFIG_DAC_LOG_LEVEL);
+
+/**
+ * https://www.analog.com/media/en/technical-documentation/data-sheets/AD5601_5611_5621.pdf
+ *
+ * The Analog Device AD5601, AD5611, and AD5621 are 8, 10, and 12 bits single channel SPI DACs.
+ * */
+
+/**
+ * The AD56x1 devices support a NORMAL mode where the output is connected to the output amplifier
+ * driven by a resistor string. These devices also support 3 power down modes in which the output is
+ * disconnected from the amplifier and is either:
+ * - connected to GND through a 1 kΩ resistor
+ * - connected to GND through a 100 kΩ resistor
+ * - Disconnected (three-state)
+ *
+ * This driver only support the normal mode to stick to the regular DAC API.
+ * */
+#define DAC_AD56X1_MODE_NORMAL                 0x0000
+#define DAC_AD56X1_MODE_POWER_DOWN_1K          0x4000
+#define DAC_AD56X1_MODE_POWER_DOWN_100K        0x8000
+#define DAC_AD56X1_MODE_POWER_DOWN_THREE_STATE 0xC000
+
+struct ad56x1_config {
+	struct spi_dt_spec bus;
+	uint8_t resolution;
+};
+
+struct ad56x1_data {
+};
+
+static int ad56x1_channel_setup(const struct device *dev, const struct dac_channel_cfg *channel_cfg)
+{
+	const struct ad56x1_config *config = dev->config;
+
+	if (channel_cfg->channel_id != 0) {
+		LOG_ERR("invalid channel %i", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != config->resolution) {
+		LOG_ERR("invalid resolution %i", channel_cfg->resolution);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->internal) {
+		LOG_ERR("Internal channels not supported");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ad56x1_write_value(const struct device *dev, uint8_t channel, uint32_t value)
+{
+	const struct ad56x1_config *config = dev->config;
+	uint8_t buffer_tx[2];
+	uint16_t command = DAC_AD56X1_MODE_NORMAL;
+
+	if (value > BIT(config->resolution) - 1) {
+		LOG_ERR("invalid value %i", value);
+		return -EINVAL;
+	}
+
+	if (channel != 0) {
+		LOG_ERR("invalid channel %i", channel);
+		return -EINVAL;
+	}
+
+	const struct spi_buf tx_buf[] = {{
+		.buf = buffer_tx,
+		.len = ARRAY_SIZE(buffer_tx),
+	}};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	command |= value << (14 - config->resolution);
+	sys_put_be16(command, buffer_tx);
+
+	LOG_DBG("sending to DAC %s command 0x%02X, (value 0x%04X, normal mode)", dev->name, command,
+		value);
+	int result = spi_write_dt(&config->bus, &tx);
```

Comment:
Please place variable definitions on the top of the block, at least when possible